server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -0,0 +1,689 @@
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.codec.vectors;
+
+import org.apache.lucene.codecs.hnsw.FlatVectorsWriter;
+import org.apache.lucene.index.FieldInfo;
+import org.apache.lucene.index.FloatVectorValues;
+import org.apache.lucene.index.MergeState;
+import org.apache.lucene.index.SegmentWriteState;
+import org.apache.lucene.index.VectorSimilarityFunction;
+import org.apache.lucene.internal.hppc.IntArrayList;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.util.InfoStream;
+import org.apache.lucene.util.VectorUtil;
+import org.apache.lucene.util.quantization.OptimizedScalarQuantizer;
+import org.elasticsearch.simdvec.ES91OSQVectorsScorer;
+import org.elasticsearch.simdvec.ESVectorUtil;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.lucene.codecs.lucene102.Lucene102BinaryQuantizedVectorsFormat.INDEX_BITS;
+import static org.apache.lucene.util.quantization.OptimizedScalarQuantizer.discretize;
+import static org.apache.lucene.util.quantization.OptimizedScalarQuantizer.packAsBinary;
+import static org.elasticsearch.index.codec.vectors.IVFVectorsFormat.IVF_VECTOR_COMPONENT;
+
+/**
+ * Default implementation of {@link IVFVectorsWriter}. It uses {@link KMeans} algorithm to
+ * partition the vector space, and then stores the centroids an posting list in a sequential
+ * fashion.
+ */
+public class DefaultIVFVectorsWriter extends IVFVectorsWriter {
+
+    static final float SOAR_LAMBDA = 1.0f;
+    // What percentage of the centroids do we do a second check on for SOAR assignment
+    static final float EXT_SOAR_LIMIT_CHECK_RATIO = 0.10f;
+
+    private final int vectorPerCluster;
+
+    public DefaultIVFVectorsWriter(SegmentWriteState state, FlatVectorsWriter rawVectorDelegate, int vectorPerCluster) throws IOException {
+        super(state, rawVectorDelegate);
+        this.vectorPerCluster = vectorPerCluster;
+    }
+
+    @Override
+    CentroidAssignmentScorer calculateAndWriteCentroids(
+        FieldInfo fieldInfo,
+        FloatVectorValues floatVectorValues,
+        IndexOutput centroidOutput,
+        float[] globalCentroid
+    ) throws IOException {
+        if (floatVectorValues.size() == 0) {
+            return CentroidAssignmentScorer.EMPTY;
+        }
+        // calculate the centroids
+        int maxNumClusters = ((floatVectorValues.size() - 1) / vectorPerCluster) + 1;
+        int desiredClusters = (int) Math.max(maxNumClusters / 16.0, Math.max(Math.sqrt(floatVectorValues.size()), maxNumClusters));
```

Comment:
It is not obvious why we divide maxNumClusters/16, a comment would help. Is it to prevent too many clusters?

Reply:
good call out. I will simplify this.